### PR TITLE
i/6536: Implemented the select all feature

### LINKED ISSUE #6536: Handle Ctrl/Cmd+A in a root
Currently, the native Ctrl+A is broken in Chrome when a widget is at the end/start of the root. If we handled this manually, then it would most likely work just fine. Let's try this, especially that we already handle Ctrl+A in nested editables.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by oleq**
Feature: Implemented the select all feature. Closes ckeditor/ckeditor5#6536.

---

To be merged together with:

*   A [**PR in ckeditor5-essentials**](https://github.com/ckeditor/ckeditor5-essentials/pull/7) making this an essential editor feature.
*   A [**PR in ckeditor5-widget**](https://github.com/ckeditor/ckeditor5-widget/pull/121) in which I removed the Ctrl+A handling added by the `Widget` plugin. `SelectAll` takes care of widgets now.
*   A [**PR in ckeditor5-theme-lark**](https://github.com/ckeditor/ckeditor5-theme-lark/pull/281) adding the select all icon to the manual test.
*   A [branch in ckeditor5](https://github.com/ckeditor/ckeditor5/compare/i/6536?expand=1).

---

Q: Do we need a feature guide? That would be a very short one but maybe it's worth adding. cc @Reinmar

---

FYI: The feature brings a UI button, which looks as follows:

image

### CODE REVIEW

File: tests/selectallcommand.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		return editor.destroy();
+	} );
+
+	describe( 'isEnabled', () => {
+		it( 'should always be "true" because the command is stateless', () => {
+			expect( command.isEnabled ).to.be.true;
+		} );
+	} );
+
+	describe( 'execute()', () => {
```
Comment (jodator) [inline tests/selectallcommand.js:39]: I think that we are safe but WDYT about adding a table test?

This would be a scenario for nested-nested editable.

**Review by jodator — COMMENTED** [review]
Basically I see room for improvement in tests. Maybe it is worth testing it with table > image also?

I'm confused with the `keystrokes` vs `listenTo( viewDocument )` usage as I don't see difference in them and personally I would use `editor.keystrokes`. So this is question to @Reinmar alongside @oleq ones.

ps.: besides that LGTM :+1: for merging.

File: tests/selectallcommand.js (hunk 1 above)
Comment (oleq) [reply tests/selectallcommand.js:39]: Nested editable is tested in the image caption.

Comment (jodator) [reply tests/selectallcommand.js:39]: Yes, but a nested editable in a nested editable case - to be sure that we select inner, not the outer nested editable.

Comment (oleq) [reply tests/selectallcommand.js:39]: Alright, makes sense 👍

**Comment by oleq** [comment]
Ready for another round of review.

**Comment by oleq** [comment]
An afterthought: do you think it would be a nice UX if pressing Ctrl+A in a nested editable which content is already selected (double Ctrl+A) selected the closest limit ancestor? 

For instance, root->table->cell->image width caption. Double Ctrl+A in the caption would select the entire content of a cell. And a double Ctrl+A would select the entire root.

**Comment by Reinmar** [comment]
> Q: Do we need a feature guide? That would be a very short one but maybe it's worth adding. cc @Reinmar

:+1:  because of the "select all"  button. Otherwise, it will be extremely hard to find.

**Comment by Reinmar** [comment]
> An afterthought: do you think it would be a nice UX if pressing Ctrl+A in a nested editable which content is already selected (double Ctrl+A) selected the closest limit ancestor?

Just to make sure that I understand correctly. In the following case:

> For instance, root->table->cell->image->caption

1.  Ctrl+A to select image caption
2.  Ctrl+A to select the image
3.  Ctrl+A to select the cell content (if the image is not already its only content)
4.  Ctrl+A to select the entire table
5.  Ctrl+A to select the entire root?

Sounds great.

**Comment by oleq** [comment]
Honestly, I’d skip 2 and 4. They don’t feel like select all but widget features.

**Comment by Reinmar** [comment]
~BTW, the way to implement that behaviour of selecting the next "thing" on Ctrl+A press when entire content of a limit element is selected is to look for the next object element. That'll be more semantically correct than looking for another limit element.~

~All objects are by default limit elements, so looking for another limit would work fine. But conceptually we're looking then for another object.~

See https://github.com/ckeditor/ckeditor5-select-all/pull/1#issuecomment-614579043.

**Comment by Reinmar** [comment]
> Honestly, I’d skip 2 and 4. They don’t feel like select all but widget features.

Could be this way too :+1:

This would also invalidate my previous comment.

File: src/selectallui.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+		editor.ui.componentFactory.add( 'selectAll', locale => {
+			const command = editor.commands.get( 'selectAll' );
+			const view = new ButtonView( locale );
+			const t = locale.t;
+
+			view.set( {
+				label: t( 'Select all' ),
+				icon: selectAllIcon,
+				keystroke: 'CTRL+A',
```
Comment (Reinmar) [inline src/selectallui.js]:
```suggestion
				keystroke: 'Ctrl+A',
```

We're not consistent here (it's wrong in basic-styles only) but that was the consensus to write Ctrl/Shift/Alt.

Comment (Reinmar) [reply src/selectallui.js]: That's also what you used later on in the manual test.

**Comment by oleq** [comment]
> 👍  because of the "select all"  button. Otherwise, it will be extremely hard to find.

Created.

> Sounds great.

I created a follow-up: https://github.com/ckeditor/ckeditor5/issues/6621. I think this could be a great intro ticket.

---

Ready for another round of review :)

**Pull request merged**